src/Features/CSharp/Portable/IntroduceUsingStatement/CSharpIntroduceUsingStatementCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,72 @@
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.IntroduceUsingStatement;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.IntroduceUsingStatement
+{
+    [ExtensionOrder(Before = PredefinedCodeRefactoringProviderNames.IntroduceVariable)]
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = PredefinedCodeRefactoringProviderNames.IntroduceUsingStatement), Shared]
+    internal sealed class CSharpIntroduceUsingStatementCodeRefactoringProvider
+        : AbstractIntroduceUsingStatementCodeRefactoringProvider<LocalDeclarationStatementSyntax>
+    {
+        protected override async Task<Document> IntroduceUsingStatementAsync(
+            Document document,
+            LocalDeclarationStatementSyntax declarationSyntax,
+            ILocalSymbol localVariable,
+            CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            if (!(declarationSyntax.Parent is BlockSyntax parent))
+            {
+                throw new NotImplementedException();
+            }
+
+            var lastStatement = (StatementSyntax)FindSiblingContainingLastUsage(
+                declarationSyntax,
+                localVariable,
+                semanticModel,
+                document.GetLanguageService<ISyntaxFactsService>(),
+                cancellationToken);
+
+            var declarationSyntaxIndex = parent.Statements.IndexOf(declarationSyntax);
+            var replaceCount = 1;
+            var block = SyntaxFactory.Block();
+
+            if (lastStatement != null)
+            {
+                var lastStatementIndex = parent.Statements.IndexOf(lastStatement, declarationSyntaxIndex + 1);
+
+                block = SyntaxFactory.Block(parent.Statements
+                    .Take(lastStatementIndex + 1)
+                    .Skip(declarationSyntaxIndex + 1));
+
+                replaceCount += lastStatementIndex - declarationSyntaxIndex;
+            }
+
+            var usingStatementSyntax =
+                SyntaxFactory.UsingStatement(
+                    declaration: declarationSyntax.Declaration,
+                    expression: null, // Declaration already has equals token and expression
+                    statement: block);
+
+            var newParent = parent
+                .WithStatements(parent.Statements
+                    .ReplaceRange(declarationSyntaxIndex, replaceCount, new[] { usingStatementSyntax }));
+
+            return document
+                .WithSyntaxRoot(root
+                    .ReplaceNode(parent, newParent));
```

Comment:
it feels like most of this logic could move down.  the overall shape would be nearly teh same for VB.  Several things could be figured out commonly with IOperatoin.  And there woudl only need to be a few points of very specialized work for C#/VB.